openmoc/materialize.py

```diff
@@ -2,192 +2,319 @@
-                'energy groups %s is not an integer', filename, str(num_groups))
-
-    data = data['Materials']
-    material_names = data.keys()
-
-    # Loop over each material and
-    for name in material_names:
-
-      py_printf('INFO', 'Importing material %s', str(name))
-
-      new_material = openmoc.Material()
-      new_material.setNumEnergyGroups(int(num_groups))
-
-      if 'Total XS' in data[name].keys():
-        new_material.setSigmaT(data[name]['Total XS'])
+    if '# groups' not in f.attrs:
+        py_printf('ERROR', 'Unable to load HDF5 file "%s" since it does '
+                           'not contain an \'# groups\' attribute', filename)
+
+    if domain_type not in f.keys():
+        py_printf('ERROR', 'Unable to load HDF5 file "%s" since it does '
+                           'not contain domain type "%s"', filename, domain_type)
+
+    if geometry and 'openmoc.Geometry' not in str(type(geometry)):
+        py_printf('ERROR', 'Unable to load HDF5 file "%s" for "%s" which is not '
+                           'an OpenMOC Geometry', filename, str(type(geometry)))
+
+    # Instantiate dictionary to hold Materials to return to user
+    materials = {}
+    num_groups = int(f.attrs['# groups'])
+
+    # If a Geometry was passed in, extract all cells or materials from it
+    if geometry:
+        if domain_type == 'material':
+            domains = geometry.getAllMaterials()
+        elif domain_type == 'cell':
+            domains = geometry.getAllMaterialCells()
+        else:
+            py_printf('ERROR', 'Domain type "%s" is not supported', domain_type)
+
+    # Iterate over all domains (e.g., materials or cells) in the HDF5 file
+    for domain_spec in f[domain_type]:
+
+        py_printf('INFO', 'Importing cross sections for %s "%s"',
+                          domain_type, str(domain_spec))
+
+        # Create shortcut to HDF5 group for this domain
+        domain_group = f[domain_type][domain_spec]
+
+        # If domain_spec is an integer, it is an ID; otherwise a string name
+        if domain_spec.isdigit():
+            domain_spec = int(domain_spec)
+        else:
+            domain_spec = str(domain_spec)
+
+        # If using an OpenMOC Geometry, extract a Material from it
+        if geometry:
+            if domain_type == 'material':
+                material = _get_domain(domains, domain_spec)
+            elif domain_type == 'cell':
+                material = _get_domain(domains, domain_spec).getFillMaterial()
+
+        # If not Geometry, instantiate a new Material with the ID/name
+        else:
+            if isinstance(domain_spec, int):
+                material = openmoc.Material(id=domain_spec)
+            else:
+                material = openmoc.Material(name=domain_spec)
+
+        # Add material to the collection
+        materials[domain_spec] = material
+        material.setNumEnergyGroups(num_groups)
+
+        # Search for the total/transport cross section
```

Comment:
I added some "INFO" log messages to `openmoc.materialize` in an attempt to help make it clear to the user which cross section types are loaded at runtime. I think our code comments on sigmaT are pretty clear, but perhaps that's not the case to a new users. Our online documentation on the `openmoc.materialize` is due for an update with both this PR and #216, and I think that this is something we can carefully note there.